mesonbuild/backend/ninjabackend.py

```diff
@@ -2006,9 +2006,10 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
         target_deps = target.get_dependencies()
         has_shared_deps = any(isinstance(dep, build.SharedLibrary) for dep in target_deps)
         if isinstance(target, build.SharedLibrary) or has_shared_deps:
-            has_rust_shared_deps = any(isinstance(dep, build.SharedLibrary) and dep.uses_rust() and dep.rust_crate_type != 'cdylib'
+            has_rust_shared_deps = any(isinstance(dep, build.SharedLibrary) and dep.uses_rust()
+                                       and dep.rust_crate_type not in ['cdylib', 'proc-macro']
                                        for dep in target_deps)
-            if cratetype != 'cdylib' or has_rust_shared_deps:
+            if cratetype not in ['cdylib', 'proc-macro'] or has_rust_shared_deps:
```

Comment:
```suggestion
            if cratetype not in {'cdylib', 'proc-macro'} or has_rust_shared_deps:
```

Reply:
I tried that in another PR and then the linter complained :) See https://github.com/mesonbuild/meson/pull/11698#issuecomment-1513037378

Reply:
The linter should complain about for loops :) because the entire point of using sets preferably instead of lists is in the context of container membership checks.
- "if x in collection" must check every element for membership but order doesn't matter. Hashed collections are faster.
- "for x in collection" must iterate over each element and utilize it in an operation, so hashed collections are irrelevant. Also order may or may not matter depending on the minutiae of the operations done, the presence of short circuiting, etc. Using sets there is either slower (wasted time hashing) or an outright bug due to the order changing unexpectedly.

Reply:
That actually makes sense, thanks :)

Reply:
(done)